### DISCUSSION TITLE
Counter callback function does not always run immediately when alarm is trigged

### DISCUSSION THREAD

**Discussion by NineDayCC** [zephyrproject-rtos/zephyr#61043] [General]
I use `counter_set_channel_alarm` to set counter alarm value and change GPIO output level in callback function.

But I found that **_the interval between the output level changes have a slightly different (around 0 - 15us)  everytime_**, which is supposed to be exactaly the same everytime. 

Is it because the kernel is doing something before callback function runs resaulting different time gap? How can I do to make the output interval run as exactly the same as the value is set? 

The counter is 1Mhz (1us). The sys clk is 160Mhz. This is part of how I setup the output.
```
//esp32c3_devkitm
&timer0 {
    status = "ok";
    label = "PPM Timer";
    prescaler = <80>;
    group = <0>;
    index = <0>;
};
```
```
ppm_init()
{
    ...
    counter_start(counter_dev);

    alarm_cfg.flags = 1;
    alarm_cfg.ticks = counter_us_to_ticks(counter_dev, DELAY);
    alarm_cfg.callback = ppm_counter_interrupt_fn;
    alarm_cfg.user_data = &alarm_cfg;

    ret = counter_set_channel_alarm(counter_dev, ALARM_CHANNEL_ID,
                                    &alarm_cfg);
    ...
}
```
```
void ppm_counter_interrupt_fn(const struct device *counter_dev,
                                     uint8_t chan_id, uint32_t ticks,
                                     void *user_data)
{
    static const struct gpio_dt_spec ppm_dev = GPIO_DT_SPEC_GET(DT_PPM_PIN, gpios);
    static uint32_t ticks_offset = 0;    //the starting_value of a new loop

    gpio_pin_toggle_dt(&ppm_dev);    //change io output level

    if (curstep == 0)
    {
        ticks_offset = ticks;
    }

    struct counter_alarm_cfg *config = user_data;
    config->ticks = ticks_offset + interval_offset[curstep];    //add next interval_offset to the starting_value of this loop as the next alarm value
    //interval_offset[] sample: {300, 600, 900, 1200}, in this case, the output gpio level should change every 300us

    curstep++;
    if (curstep > (sizeof(interval) / sizeof(interval[0]) ) )
    {
        curstep = 0;
    }

    counter_set_channel_alarm(counter_dev, ALARM_CHANNEL_ID, user_data);
    ...
}
```

Information may help:
1. The **minimum interval** I run between level changes is **300us**. When I use logic analyzer to read the output, it's not alaways accurate 300us.
2. The application also runs **BT, SPI** in other threads.

OS: zephyr 3.4.0
Toolchain: zephyr-sdk-0.16.0
Board: esp32c3_devkitm

**Comment by nordic-krch** [comment]
Since CPU toggles the gpio and not the hardware there will always be a jitter because there are other interrupts, cache, etc. As you mentioned, your application also runs BT so you have interrupts from radio. You can try to increase interrupt priority as it's likely that by default all interrupts have the same priority.
